File: app/routers/reports.py
from fastapi import APIRouter, Depends, Request
from fastapi.responses import HTMLResponse, StreamingResponse
from sqlalchemy.orm import Session
from app.database import get_db
from app.deps import require_admin
from app.templating import render
from app import models
import io
import csv

router = APIRouter(prefix="/admin", dependencies=[Depends(require_admin)])

@router.get("/pit", response_class=HTMLResponse)
def pit_uke_report_page(request: Request, db: Session = Depends(get_db)):
    from sqlalchemy import func, select
    nodes_count = db.scalar(select(func.count()).select_from(models.CustomerDevice)) or 0
    return render(request, "admin/pit_uke.html", {
        "title": "Eksport PIT UKE",
        "nodes_count": nodes_count
    })

@router.get("/pit/export")
def pit_uke_export_csv(db: Session = Depends(get_db)):
    from sqlalchemy import select
    from sqlalchemy.orm import joinedload
    output = io.StringIO()
    writer = csv.writer(output, delimiter=';')
    writer.writerow(['ID', 'IP', 'MAC', 'Customer', 'Street', 'Number'])
    
    stmt = (
        select(models.CustomerDevice)
        .options(
            joinedload(models.CustomerDevice.customer)
            .joinedload(models.Customer.street)
        )
    )
    nodes = db.scalars(stmt).all()
    
    for n in nodes:
        cust = n.customer
        cust_name = f"{cust.first_name} {cust.last_name}" if cust else "—"
        street_name = cust.street.name if (cust and cust.street) else ""
        street_num = cust.street_number if cust else ""
        
        writer.writerow([
            n.id, 
            n.ip_address or "", 
            n.mac_address or "", 
            cust_name,
            street_name,
            street_num
        ])
    
    output.seek(0)
    return StreamingResponse(
        io.BytesIO(output.getvalue().encode('utf-8-sig')),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=pit_uke_export.csv"}
    )

@router.get("/passport/map", response_class=HTMLResponse)
def network_map_page(request: Request, db: Session = Depends(get_db)):
    from sqlalchemy import select
    from sqlalchemy.orm import joinedload
    # Fetch nodes with coordinates
    stmt = select(models.NetNode).options(joinedload(models.NetNode.location_city))
    nodes = db.scalars(stmt).all()
    # Convert to simple list for JSON
    nodes_data = []
    for n in nodes:
        if n.latitude and n.longitude:
            nodes_data.append({
                "name": n.name,
                "lat": float(n.latitude),
                "lon": float(n.longitude),
                "type": "net_node",
                "address": f"{n.location_city.name if n.location_city else ''} {n.street_number or ''}"
            })
            
    import json
    return render(request, "net_map.html", {
        "title": "Mapa Sieci",
        "nodes_json": json.dumps(nodes_data)
    })

File: crm-portal-ts/src/routers/reports.ts
import { Router } from "express";
import { AppDataSource } from "../database.js";
import { CustomerDevice, NetNode } from "../models/network.js";
import { LocationStreet } from "../models/location.js";

export const router = Router();

const customerDeviceRepo = AppDataSource.getRepository(CustomerDevice);
const netNodeRepo = AppDataSource.getRepository(NetNode);
const locationStreetRepo = AppDataSource.getRepository(LocationStreet);

function escapeCsv(value: unknown) {
    const text = String(value ?? "");
    return `"${text.replaceAll("\"", "\"\"")}"`;
}

router.get("/pit-uke/summary", async (_req, res) => {
    try {
        const customerDeviceCount = await customerDeviceRepo.count();
        res.json({ customerDeviceCount });
    } catch (error) {
        console.error("Error building PIT UKE summary:", error);
        res.status(500).json({ message: "Internal server error" });
    }
});

router.get("/pit-uke/export", async (_req, res) => {
    try {
        const devices = await customerDeviceRepo.find({
            relations: {
                customer: true,
            },
            order: { id: "ASC" },
        });

        const streetIds = devices
            .map((device) => device.customer?.locationStreetId)
            .filter((id): id is number => Number.isInteger(id));
        const streets = streetIds.length > 0
            ? await locationStreetRepo.findByIds(streetIds)
            : [];
        const streetMap = new Map(streets.map((street) => [street.id, street.name]));

        const rows = [
            ["ID", "IP", "MAC", "Customer", "Street", "Number"],
            ...devices.map((device) => {
                const customer = device.customer;
                const customerName = customer
                    ? `${customer.firstName} ${customer.lastName}`.trim()
                    : "";
                const streetName = customer?.locationStreetId
                    ? streetMap.get(customer.locationStreetId) ?? ""
                    : "";

                return [
                    device.id,
                    device.ipAddress ?? "",
                    device.macAddress ?? "",
                    customerName,
                    streetName,
                    customer?.streetNumber ?? "",
                ];
            }),
        ];

        const csv = `${rows
            .map((row) => row.map((value) => escapeCsv(value)).join(";"))
            .join("\r\n")}\r\n`;

        res.setHeader("Content-Type", "text/csv; charset=utf-8");
        res.setHeader("Content-Disposition", "attachment; filename=\"pit_uke_export.csv\"");
        res.send(`\uFEFF${csv}`);
    } catch (error) {
        console.error("Error exporting PIT UKE CSV:", error);
        res.status(500).json({ message: "Internal server error" });
    }
});

router.get("/passport/map", async (_req, res) => {
    try {
        const nodes = await netNodeRepo.find({ order: { name: "ASC" } });
        const nodesData = nodes
            .filter((node) => Number.isFinite(node.latitude) && Number.isFinite(node.longitude))
            .map((node) => ({
                id: node.id,
                name: node.name,
                lat: node.latitude,
                lon: node.longitude,
                type: "net_node",
                address: node.locationDetail ?? "",
            }));

        res.json(nodesData);
    } catch (error) {
        console.error("Error building passport map payload:", error);
        res.status(500).json({ message: "Internal server error" });
    }
});
